Sign in with Klarna

mntzrr opened the pull request:
Make sure you're using a whitelisted URL if you need to test it.

Related task: https://app.clickup.com/t/860qqd1aa

mntzrr committed:
Create new user with Klarna

mntzrr committed:
Handle expired refresh token

mntzrr committed:
Save refresh token if already logged in

mntzrr committed:
Add plugin settings

mntzrr committed:
Move class-siwk-ajax.php to Ajax.php

mntzrr committed:
Add setting to set cart page placement

mntzrr committed:
Move settings to new Settings class

mntzrr committed:
Move user management to new User class

MichaelBengtsson commented on src/Ajax.php:
```diff
@@ -0,0 +1,157 @@
+	exit;
+}
+
+require_once 'includes/functions.php';
```
You should not have a require_once like this. Id rather see a class that handles the JWT stuff in this case

mntzrr committed:
Move JWT functions to new JWT class

MichaelBengtsson commented on src/SignInWithKlarna.php:
```diff
@@ -0,0 +1,119 @@
+			return;
+		}
+
+		$path        = dirname( __FILE__ ) . '/assets/siwk.js';
+		$script_path = substr( $path, strpos( $path, '/wp-content' ) );
```
Id like to discuss this if possible. I just want to make sure this works as intended and wont cause issues.

Getting the URL for the assets is probably one of the most complicated things when it comes to doing this in the package, so i just want to go over possible issues if we can.

MichaelBengtsson reviewed (commented):
It looks good, i just want to go over the assets url as i mentioned in my other comment before we make a release. Lets discuss it on the standup today and plan a meeting about it.

mntzrr committed:
[PSR-4 compliance] Rename Ajax to AJAX

mntzrr committed:
Add region setting and set default values

mntzrr committed:
Add JWT validation and remove use of static

mntzrr committed:
Refactor static function to non-static and improve error handling

mntzrr committed:
Do not extend WC_AJAX and convert static functions to non-static

mntzrr committed:
Refactor to use non-static classes

mntzrr committed:
Use WP_Error for passing along error messages

mntzrr committed:
Use asset relative path to plugin directory

mntzrr committed:
Fix typo in function name

mntzrr committed:
Add callback page

mntzrr committed:
Identity button not needed on callback page

mntzrr committed:
Move scope to Klarna button

mntzrr committed:
Manage all things JWT related in JWT class

mntzrr committed:
Remove unused parameters

mntzrr committed:
Run placement hook only once without removing action

mntzrr committed:
Register API endpoint for handling callback from Klarna

mntzrr committed:
Extract user management to User class

mntzrr committed:
Add support for region NA

mntzrr committed:
rest_param not needed

WC require pretty permalinks therefore we can assume it is enabled.

mntzrr committed:
Move shared settings to internal settings

mntzrr committed:
Handle sign in from callback page

mntzrr committed:
Redirect customer back to account page

mntzrr committed:
popout is popup

mntzrr committed:
Redirect the customer to the shop

mntzrr committed:
Add new button theme

mntzrr committed:
Add new button shape

mntzrr committed:
Replace environment with test mode

mntzrr committed:
Update badge settings to reflect new values

mntzrr committed:
Changed setting description

mntzrr committed:
Add siwk_redirect_url filter

mntzrr committed:
Add button to mini cart

mntzrr committed:
Default to empty array if billing address is not granted

mntzrr committed:
Display the sign-in button at the start of my-account page

mntzrr committed:
Handle callback for redirect flow in Redirect class

mntzrr committed:
Show callback URL in settings page

mntzrr committed:
Use uniform endpoint for handling redirect callback

To remain consistent across platforms, the endpoint has been changed to `siwk/klarna/callback`.

mntzrr committed:
Display the current scopes on the settings page

mntzrr committed:
badge_alignment is logo_alignment

mntzrr committed:
Move filters to Setting

mntzrr committed:
Ensure SIWK_VERSION is defined only once

Since the code can be executed multiple times, the `SIWK_VERSION` will be continuously defined.

mntzrr committed:
Extend the new KP settings page

mntzrr committed:
Retrieve option for test mode from KP

mntzrr committed:
Retrieve client ID and region from KP

mntzrr committed:
Fix missing attributes and remove magic method

mntzrr committed:
Update scoping

Not publicly documented yet but `payment:request:create` is replaced by `customer:login`.

mntzrr committed:
Merchant can set scope from settings page

mntzrr committed:
Hard-code the callback URL

mntzrr committed:
Match label naming in MePo

mntzrr committed:
Add copy-to-clipboard

mntzrr committed:
Adjust absolute positioning

mntzrr committed:
Default to 'no'

mntzrr committed:
Make billing address scope required

mntzrr committed:
Adds a preview on the settings page

mntzrr committed:
Store the preview button images locally

mntzrr committed:
Add JS linting

mntzrr committed:
Update the preview image on select change

mntzrr committed:
Remove unused function

mntzrr committed:
Default to empty string when data is missing from scope

mntzrr committed:
Explicitly add scopes for required data

mntzrr committed:
Fixes country check for client ID always picking EU

mntzrr committed:
Make public property private

mntzrr committed:
Checks if country should be available before retrieving client ID

mntzrr committed:
Adds class for targeting the SIWK button(s)

mntzrr committed:
Maximizes SIWK button to occupy all container space

mntzrr committed:
Use same logic for outputting button

mntzrr committed:
Only enqueues if the button should be displayed

mntzrr committed:
Redirects customer back to shop page on sign in

mntzrr committed:
Adds instance name to resolve Web SDK conflicts

> "Another Klarna Web SDK was already included into the page, please consider using data-client-instance-name attribute on klarna.js script tag to avoid conflicts."

I suspect this happening because KEC, KOSM and SIWK are enqueuing the same library.

mntzrr committed:
Customer will remain on cart page on sign in

mntzrr committed:
Removes data-client-instance-name

mntzrr committed:
Passes the $page where available to redirect URL filter

mntzrr committed:
Sanitizes the current page URL received from the frontend

mntzrr committed:
Retrieve the client ID using KP helper function

mntzrr committed:
Prefix hook name to avoid collision

mntzrr committed:
Prefix the callback function with 'siwk'

mntzrr committed:
Do not overwrite shipping address if it exists

MichaelBengtsson committed:
Merge pull request #2 from krokedil/develop-websdk

Migrate to the new WebSDK
